### PR TITLE
V3 api refactor

### PR DESCRIPTION
**Pull request by mythmon**
I think this makes the API a lot easier to understand. Previously, there was a confusing mix of hierarchy and recursion. When querying `/v3/api/recipe/{id}/`, some fields were placed at the top level, and other fields were placed within a latest or approved revision key, which then had a recipe key that duplicated those top level keys again.

The goal here is not to minimize bytes-on-the-wire, but to make the overall structure clearer and better reflect the underlying data. As a bonus, enabled states are now exposed in the API.

### CODE REVIEW

**Comment by mythmon** [comment]
@peterbe I think Rehan is the right person to review this, but I am also interested in your thoughts about this.

Here is an example of the output of API v2 and v3 (where v2 is very similar to what v3 used to be):

# Before (v2)

`$ curl -k https://localhost:8000/api/v2/recipe/1/ | jq`
<details>
<pre><code>
{
  "action": {
    "arguments_schema": {
      "$schema": "http://json-schema.org/draft-04/schema#",
      "properties": {
        "message": {
          "default": "",
          "description": "Message to log to the console",
          "type": "string"
        }
      },
      "required": [
        "message"
      ],
      "title": "Log a message to the console",
      "type": "object"
    },
    "id": 3,
    "implementation_url": null,
    "name": "console-log"
  },
  "approval_request": null,
  "approved_revision": {
    "approval_request": {
      "approved": true,
      "approver": {
        "email": "mcooper@mozilla.com",
        "first_name": "Michael",
        "id": 1,
        "last_name": "Cooper"
      },
      "comment": "r+",
      "created": "2018-09-19T21:55:41.238416Z",
      "creator": {
        "email": "mcooper@mozilla.com",
        "first_name": "Michael",
        "id": 1,
        "last_name": "Cooper"
      },
      "id": 1
    },
    "comment": "",
    "date_created": "2018-08-23T22:59:17.561775Z",
    "id": 1,
    "identicon_seed": "v1:m4leny2mgx9",
    "recipe": {
      "action": {
        "arguments_schema": {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": {
            "message": {
              "default": "",
              "description": "Message to log to the console",
              "type": "string"
            }
          },
          "required": [
            "message"
          ],
          "title": "Log a message to the console",
          "type": "object"
        },
        "id": 3,
        "implementation_url": null,
        "name": "console-log"
      },
      "arguments": {
        "message": "asasfdadf"
      },
      "enabled": false,
      "extra_filter_expression": "true",
      "filter_expression": "true",
      "filter_object": [],
      "id": 1,
      "identicon_seed": "v1:m4leny2mgx9",
      "is_approved": true,
      "last_updated": "2018-09-19T21:57:52.610463Z",
      "name": "test 2018-08-23 - a"
    },
    "user": null
  },
  "arguments": {
    "message": "asasfdadf"
  },
  "enabled": false,
  "extra_filter_expression": "true",
  "filter_expression": "true",
  "filter_object": [],
  "id": 1,
  "identicon_seed": "v1:m4leny2mgx9",
  "is_approved": true,
  "last_updated": "2018-09-19T21:57:52.610463Z",
  "latest_revision": {
    "approval_request": null,
    "comment": "",
    "date_created": "2018-09-19T22:01:10.148723Z",
    "id": 2,
    "identicon_seed": "v1:m4leny2mgx9",
    "recipe": {
      "action": {
        "arguments_schema": {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": {
            "message": {
              "default": "",
              "description": "Message to log to the console",
              "type": "string"
            }
          },
          "required": [
            "message"
          ],
          "title": "Log a message to the console",
          "type": "object"
        },
        "id": 3,
        "implementation_url": null,
        "name": "console-log"
      },
      "arguments": {
        "message": "asasfdadf + 1"
      },
      "enabled": false,
      "extra_filter_expression": "true",
      "filter_expression": "true",
      "filter_object": [],
      "id": 1,
      "identicon_seed": "v1:m4leny2mgx9",
      "is_approved": false,
      "last_updated": "2018-09-19T22:01:10.149889Z",
      "name": "test 2018-08-23 - a"
    },
    "user": {
      "email": "mcooper@mozilla.com",
      "first_name": "Michael",
      "id": 1,
      "last_name": "Cooper"
    }
  },
  "name": "test 2018-08-23 - a"
}
</code></pre>
</details>

# After (new v3)

`$ curl -k https://localhost:8000/api/v3/recipe/1/ | jq`
<details>
<pre><code>
{
  "approved_revision": {
    "action": {
      "arguments_schema": {
        "$schema": "http://json-schema.org/draft-04/schema#",
        "properties": {
          "message": {
            "default": "",
            "description": "Message to log to the console",
            "type": "string"
          }
        },
        "required": [
          "message"
        ],
        "title": "Log a message to the console",
        "type": "object"
      },
      "id": 3,
      "implementation_url": null,
      "name": "console-log"
    },
    "approval_request": {
      "approved": true,
      "approver": {
        "email": "mcooper@mozilla.com",
        "first_name": "Michael",
        "id": 1,
        "last_name": "Cooper"
      },
      "comment": "r+",
      "created": "2018-09-19T21:55:41.238416Z",
      "creator": {
        "email": "mcooper@mozilla.com",
        "first_name": "Michael",
        "id": 1,
        "last_name": "Cooper"
      },
      "id": 1
    },
    "arguments": {
      "message": "asasfdadf"
    },
    "bug_number": null,
    "comment": "",
    "creator": null,
    "date_created": "2018-08-23T22:59:17.561775Z",
    "enabled": false,
    "enabled_states": [
      {
        "carryover_from": null,
        "created": "2018-09-19T21:57:52.593576Z",
        "creator": {
          "email": "mcooper@mozilla.com",
          "first_name": "Michael",
          "id": 1,
          "last_name": "Cooper"
        },
        "enabled": false
      },
      {
        "carryover_from": null,
        "created": "2018-09-19T21:57:42.218659Z",
        "creator": {
          "email": "mcooper@mozilla.com",
          "first_name": "Michael",
          "id": 1,
          "last_name": "Cooper"
        },
        "enabled": true
      }
    ],
    "extra_filter_expression": "true",
    "filter_expression": "true",
    "filter_object": [],
    "id": 1,
    "identicon_seed": "v1:m4leny2mgx9",
    "name": "test 2018-08-23 - a",
    "recipe": {
      "approved_revision_id": 1,
      "id": 1,
      "latest_revision_id": 2
    },
    "updated": "2018-09-19T21:57:52.610463Z"
  },
  "id": 1,
  "latest_revision": {
    "action": {
      "arguments_schema": {
        "$schema": "http://json-schema.org/draft-04/schema#",
        "properties": {
          "message": {
            "default": "",
            "description": "Message to log to the console",
            "type": "string"
          }
        },
        "required": [
          "message"
        ],
        "title": "Log a message to the console",
        "type": "object"
      },
      "id": 3,
      "implementation_url": null,
      "name": "console-log"
    },
    "approval_request": null,
    "arguments": {
      "message": "asasfdadf + 1"
    },
    "bug_number": null,
    "comment": "",
    "creator": {
      "email": "mcooper@mozilla.com",
      "first_name": "Michael",
      "id": 1,
      "last_name": "Cooper"
    },
    "date_created": "2018-09-19T22:01:10.148723Z",
    "enabled": false,
    "enabled_states": [],
    "extra_filter_expression": "true",
    "filter_expression": "true",
    "filter_object": [],
    "id": 2,
    "identicon_seed": "v1:m4leny2mgx9",
    "name": "test 2018-08-23 - a",
    "recipe": {
      "approved_revision_id": 1,
      "id": 1,
      "latest_revision_id": 2
    },
    "updated": "2018-09-19T22:01:10.149889Z"
  },
  "signature": null
}
</code></pre>
</details>

File: normandy/recipes/api/v3/views.py [hunk 1]
```diff
@@ -137,6 +144,7 @@ def enable(self, request, pk=None):
                 status=status.HTTP_409_CONFLICT,
             )
 
+        recipe.latest_revision.refresh_from_db()
```
Comment (peterbe) [inline normandy/recipes/api/v3/views.py]: nice!

**Comment by mythmon** [comment]
> The new API now has approved_revision and latest_revision whereas before it only had latest_revision. Right?

No, the previous iteration ended up with `approved_revision`, `latest_revision`, *and* a weird mess of fields at the top level that represented some unclear mix of the two.

> And the old latest_revision only had the boolean enabled whereas the new one, has enabled_states in the approved_revision. Right?

That's about right. Both the latest and approved revisions include the enabled states field. I was a bit naughty and included this change in the refactor. Actually the addition of `enabled_states` is the primary reason that I wanted to do this refactor. This is an absolutely huge thing will close the loop on one of highest ranking requests I get about the API. (the other big one being filter objects)

> The new API has a new block for signature (in both approved_revision and latest_revision) which is new. Right?

This is new. I'm not sure how useful it is, since the signed data isn't apparent from this API response. With some hindsight I think we should either remove the signature field fro this response, or add in some extra details to make it useful. Either way, this is a bit aspirational and doesn't have any use cases yet.

> One thing that I'm failing to see clearly is why we're adding all these new pieces of data. This endpoint's first and main (and only?) "customer" it the recipe details page on Delivery Console and I don't see where it says that we need these. I'm a minimalist and would probably start from DC and work backwards instead of trying to "shove everthing in" that might be used.

First, the only new data (to my knowledge) is `enabled_states` and `signature`. Signature is dubiously useful, and might get removed. Enabled states are very very useful: without them it is actually not possible to say with certainty when a recipe was enabled or disabled. It's just not possible, and this has been a huge problem with the API for a long time. Most of the changes here are making the existing data better organized, to make it clearer where it belongs and how its nested.

Although no one else writes to the Normandy API, we have several users that read from it, including [the Normandy devtools](https://github.com/mozilla/normany-devtools), [this dashboard that S&I made about Shield studies](https://strategy-and-insights.mozilla.com/shield-studies/index.html), data analysis like Sunah Suh and William Beard who are monitoring and analyzing studies, and often times me. I query the API directly several times a day. Partially because jq is way faster than a web browser, but also because DC is awkward or impossible to use for some of the things I want to do (all of which have issues on file in DC).

> Does this mean we can now serve the History timeline widget from this one recipe GET?

No, the timeline widget needs the full recipe history, which comes from `/api/v3/recipe/{id}/history/`. The latest and approved revisions are only two in a large list of possible revisions. I'm not sure if `/api/v3/recipe/{id}/` and `/api/v3/recipe/{id}/history/` need to stay separate, but right now I'm leaning towards keeping them apart.

> Sorry for bringing more questions and constructive pointers to this.

This is actually exactly the kind of feedback I was looking for. I'm taking advantage of your relative ignorance of Normandy and its history, and trying to get an outsider's point of view while it lasts. :smile:

File: normandy/recipes/api/v3/serializers.py [hunk 2]
```diff
@@ -252,3 +294,8 @@ def validate_filter_object(self, value):
             raise serializers.ValidationError(errors)
 
         return value
+
+
+class RecipeLinkSerializer(RecipeSerializer):
```
Comment (rehandalal) [inline normandy/recipes/api/v3/serializers.py]: 👍

File: normandy/recipes/models.py [hunk 3]
```diff
@@ -276,8 +276,13 @@ def revise(self, force=False, **data):
 
     @transaction.atomic
     def save(self, *args, **kwargs):
-        if self.is_dirty(check_relationship=True):
-            dirty_fields = self.get_dirty_fields(check_relationship=True)
+        dirty_fields = {
+            k: v
+            for k, v in self.get_dirty_fields(check_relationship=True, verbose=True).items()
+            if v["saved"] != v["current"]
```
Comment (rehandalal) [inline normandy/recipes/models.py]: TIL

**Comment by mythmon** [comment]
bors r= @rehandalal

**Pull request merged**